# 01 - Data Cleaning and Preparation

**Dissertation Title:**
Engagement Trajectory Profiling in AI-Powered Personalised Storytelling for Individuals with Autism Spectrum Disorder

**Purpose**
This notebook imports the raw dataset, performs data cleaning, handles missing values, prepares the engagement variables, and saves a clean dataset for subsequent analyses.

## 2. Import Libraries

In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Display settings
pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 100)

## 2. Load the Raw Dataset

In [4]:
from pathlib import Path

# Project paths
project_root = Path.cwd().parent
raw_data_folder = project_root / "data" / "raw"

# Show the files inside the raw data folder
list(raw_data_folder.iterdir())

[WindowsPath('c:/Users/puspi/OneDrive/Desktop/Dissertation/data/raw/data_numeric_format_data.csv')]

## 2. Import the Dataset

In [5]:
# Load the dataset
df = pd.read_csv(
    raw_data_folder / "data_numeric_format_data.csv",
    encoding="cp1252"
)

# Display first 5 rows
df.head()

,Participant id,Age,Gender,Special_interest,Any co-existing disabiltiy diagnosis,Co-existing disability diagnosed details,Level of Severity,Autism Level,Stimming behaviour Identified?,Observed_stimming,Primary Goal,Other information eg anymedicalcondtion selective autism,Session number,Session date,Submitted_by,therapist_parent_name,How engaged was the participant during today's storytelling session?,Comment_Q1,How well did the participant understand that this story is personalised for them?,Unnamed: 19,Comment_Q2,Did the participant demonstrate emotional connection?,Comment_Q3,How would you rate the participant’s verbal participation?,Comment_Q4,Did the participant maintain attention throughout the session?,Comment_Q5,"How likely a participant was to retell and use the stories in his/her daily conversation or interaction with\ntheir peers, carers, parents or therapists?",Comment_Q6,Did the participant show any signs of enjoyment during the session?,Comment_Q7,"Did the participant exhibit distress, boredom, or frustration?",Comment_Q8,Did the participant initiate interactions related to the story?,Comment_Q9,Did the participant repeat the story or any content from the story?,Comment_Q10,Did the participant want to or try to creatively change the story?,Comment_Q11,How did retelling the stories at home impact the carer/participant relationship?,Comment_Q12,How much did the relationship between a participant and carer/parent improved?,Comment_Q13,Did the participant express their feelings about the story (verbally or otherwise)?,Comment_Q14,What is the response time of the participant? (in minutes roughly),Comment_Q15,Did the response time decrease from the last session?,Comment_Q16,Did the response time increase from the last session?,Comment_Q17,Theme of Today's Story,To what extent did the participant understand the story theme?,Comment_Q18,Did the participant apply the learning during the session?,Comment_Q20,Do you feel that the participant is confident and has the potential to be able to apply this story in the real world after a session?,Comment_Q21,Did the participant generalise the behaviour outside the story?,Comment_Q22,Did the participant recall or refer to a previous story or theme?,Comment_Q23,Did the participant reflect on or comment about the theme after the story ended?,Comment_Q24,Did the participant link the story to real-life experiences?,Comment_Q25,How much different scenarios stories impact overall social behaviour ?,Comment_Q26,Theme_specific_situation,Engagement_opportunities_count,Successful/postive response,Success_percentage,Notes_intervention,Additional_notes_observations
0,101,10,Male,"Shopping at the supermarket, going to the gas ...",Yes,"He has ADHD,which is currently being treated w...",6,2,Yes,Repetive questioning and repetive speecch,NaN,NaN,2,NaN,T,Asad,1.0,NaN,1.0,NaN,NaN,1.0,NaN,3.0,NaN,1.0,NaN,1.0,NaN,1.0,NaN,1.0,NaN,1.0,NaN,1.0,NaN,1.0,NaN,NaN,NaN,1.0,NaN,1.0,NaN,NaN,NaN,No,NaN,No,NaN,NaN,1.0,NaN,1.0,NaN,1.0,NaN,1.0,NaN,1.0,NaN,1.0,NaN,1.0,NaN,3,NaN,No Translation,10,3,30%,NaN,The participant demonstrated very good underst...
1,101,10,Male,"Shopping at the supermarket, going to the gas ...",Yes,"He has ADHD,which is currently being treated w...",6,2,Yes,Repetive questioning and repetive speecch,NaN,NaN,3,NaN,T,Asad,3.0,NaN,3.0,NaN,NaN,2.0,NaN,4.0,NaN,3.0,NaN,1.0,NaN,2.0,NaN,0.0,NaN,0.0,NaN,1.0,NaN,0.0,NaN,Stayed same,NaN,3.0,NaN,0.0,NaN,NaN,NaN,No,NaN,Yes,NaN,NaN,3.0,NaN,2.0,NaN,1.0,NaN,2.0,NaN,2.0,NaN,1.0,NaN,1.0,NaN,4,NaN,No Translation,10,5,50%,NaN,The participant showed enthusiasm for listenin...
2,101,10,Male,"Shopping at the supermarket, going to the gas ...",Yes,"He has ADHD,which is currently being treated w...",6,2,Yes,Repetive questioning and repetive speecch,NaN,NaN,4,NaN,T,Asad,2.0,NaN,2.0,NaN,NaN,2.0,NaN,4.0,NaN,2.0,NaN,2.0,NaN,3.0,NaN,1.0,NaN,1.0,NaN,3.0,NaN,0.0,NaN,NaN,NaN,3.0,NaN,1.0,NaN,NaN,NaN,No,NaN,Yes,NaN,NaN,1.0,NaN,2.0,NaN,1.0,NaN,2.0,NaN,2.0,NaN,2.0,NaN,2.0,NaN,5,N

In [6]:
df.shape

(512, 74)

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 512 entries, 0 to 511
Data columns (total 74 columns):
 #   Column                                                                                                                                                    Non-Null Count  Dtype  
---  ------                                                                                                                                                    --------------  -----  
 0   Participant id                                                                                                                                            512 non-null    int64  
 1   Age                                                                                                                                                       512 non-null    int64  
 2   Gender                                                                                                                                                    512 

In [8]:
df.describe()


,Participant id,Age,Level of Severity,Autism Level,Session number,How engaged was the participant during today's storytelling session?,How well did the participant understand that this story is personalised for them?,Unnamed: 19,Did the participant demonstrate emotional connection?,How would you rate the participant’s verbal participation?,Did the participant maintain attention throughout the session?,"How likely a participant was to retell and use the stories in his/her daily conversation or interaction with\ntheir peers, carers, parents or therapists?",Did the participant show any signs of enjoyment during the session?,"Did the participant exhibit distress, boredom, or frustration?",Did the participant initiate interactions related to the story?,Did the participant repeat the story or any content from the story?,Did the participant want to or try to creatively change the story?,How much did the relationship between a participant and carer/parent improved?,Did the participant express their feelings about the story (verbally or otherwise)?,To what extent did the participant understand the story theme?,Did the participant apply the learning during the session?,Do you feel that the participant is confident and has the potential to be able to apply this story in the real world after a session?,Did the participant generalise the behaviour outside the story?,Did the participant recall or refer to a previous story or theme?,Did the participant reflect on or comment about the theme after the story ended?,Did the participant link the story to real-life experiences?
count,512.000000,512.000000,512.000000,512.000000,512.000000,493.000000,487.000000,0.0,493.000000,490.000000,491.000000,490.000000,489.000000,489.000000,489.000000,490.000000,488.000000,457.000000,486.000000,488.000000,483.000000,482.000000,484.000000,468.000000,484.000000,486.000000
mean,118.843750,14.843750,4.312500,1.562500,5.500000,2.916836,3.065708,NaN,2.667343,7.193878,2.859470,2.232653,2.805726,0.950920,2.092025,2.389796,1.325820,1.803063,2.183128,2.967213,2.134576,2.244813,1.925620,2.384615,1.991736,1.944444
std,10.296913,5.356106,2.617867,0.496564,2.293529,1.037498,1.046154,NaN,1.073938,2.581427,0.995196,1.121894,1.018823,1.120617,1.220436,1.186624,1.290619,1.143513,1.182956,1.008664,1.193214,1.108568,1.148698,1.325721,1.183012,1.153658
min,101.000000,8.000000,1.000000,1.000000,2.000000,0.000000,0.000000,NaN,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,110.750000,10.000000,2.000000,1.000000,3.750000,2.000000,2.000000,NaN,2.000000,6.000000,2.000000,2.000000,2.000000,0.000000,1.000000,2.000000,0.000000,1.000000,1.250000,3.000000,1.000000,2.000000,1.000000,2.000000,1.000000,1.000000
50%,118.500000,13.000000,5.000000,2.000000,5.500000,3.000000,3.000000,NaN,3.000000,8.000000,3.000000,2.000000,3.000000,1.000000,2.000000,3.000000,1.000000,2.000000,2.000000,3.000000,2.000000,2.000000,2.000000,3.000000,2.000000,2.000000
75%,127.250000,18.250000,7.000000,2.000000,7.250000,4.000000,4.000000,NaN,3.000000,9.000000,4.000000,3.000000,3.000000,2.000000,3.000000,3.000000,2.000000,3.000000,3.000000,4.000000,3.000000,3.000000,3.000000,3.000000,3.000000,3.000000
max,136.000000,26.000000,8.000000,2.000000,9.000000,4.000000,4.000000,NaN,4.000000,10.000000,4.000000,4.000000,4.000000,4.000000,4.000000,4.000000,4.000000,4.000000,4.000000,4.000000,4.000000,4.000000,4.000000,4.000000,4.000000,4.000000


## Missing Value Analysis

In [9]:
df.isnull().sum()

Participant id                                                                                                                                                 0
Age                                                                                                                                                            0
Gender                                                                                                                                                         0
Special_interest                                                                                                                                               0
Any co-existing disabiltiy diagnosis                                                                                                                           0
Co-existing disability diagnosed details                                                                                                                     112
Level of Severity                 

In [10]:
(df.isnull().sum() / len(df)) * 100

Participant id                                                                                                                                                 0.000000
Age                                                                                                                                                            0.000000
Gender                                                                                                                                                         0.000000
Special_interest                                                                                                                                               0.000000
Any co-existing disabiltiy diagnosis                                                                                                                           0.000000
Co-existing disability diagnosed details                                                                                                                      21

### Interpretation

The percentage of missing values was calculated for each variable in the dataset.

Initial inspection indicates that:

- Several participant demographic variables contain no missing values.
- Most quantitative engagement rating variables contain relatively few missing values.
- Several comment fields contain a high percentage of missing values because qualitative comments were optional rather than mandatory.
- Some administrative variables (e.g., unnamed columns) contain no useful information and may be removed during preprocessing.

The missing value percentages will be used to determine an appropriate strategy for retaining, cleaning, or removing variables before further analysis.

To facilitate interpretation, missing value percentages were calculated and sorted in descending order.

In [11]:
missing_percentage = (df.isnull().sum() / len(df)) * 100
missing_percentage.sort_values(ascending=False)

Unnamed: 19                                                                                                                                                  100.000000
Comment_Q26                                                                                                                                                   99.414062
Comment_Q12                                                                                                                                                   99.218750
Comment_Q23                                                                                                                                                   98.632812
Comment_Q20                                                                                                                                                   98.242188
Comment_Q2                                                                                                                                                    98

### Interpretation

The missing value analysis revealed considerable variation in data completeness across variables.

Several participant demographic and quantitative rating variables contained little or no missing data, making them suitable for subsequent analysis.

Conversely, many qualitative comment fields contained over 90% missing values, suggesting that narrative comments were recorded only when therapists considered additional observations necessary rather than as mandatory responses.

One unnamed column contained 100% missing values, indicating that it provides no analytical value and should be removed during preprocessing.

These findings will guide feature selection and preprocessing decisions before clustering and NLP analysis.

# Feature Inspection

Before removing or transforming any variables, the dataset columns are inspected to understand their names, structure, and relevance to the research objectives. This ensures that preprocessing decisions are transparent, justified, and reproducible.

In [12]:
df.columns

Index(['Participant id', 'Age', 'Gender', 'Special_interest',
       'Any co-existing disabiltiy diagnosis',
       'Co-existing disability diagnosed details', 'Level of Severity',
       'Autism Level', 'Stimming behaviour Identified?', 'Observed_stimming',
       'Primary Goal',
       'Other information eg anymedicalcondtion selective autism',
       'Session number', 'Session date', 'Submitted_by',
       'therapist_parent_name',
       'How engaged was the participant during today's storytelling session?',
       'Comment_Q1',
       'How well did the participant understand that this story is personalised for them?',
       'Unnamed: 19', 'Comment_Q2',
       'Did the participant demonstrate emotional connection?', 'Comment_Q3',
       'How would you rate the participant’s verbal participation?',
       'Comment_Q4',
       'Did the participant maintain attention throughout the session?',
       'Comment_Q5',
       'How likely a participant was to retell and use the stories in hi

## Feature Categorisation

The dataset contains demographic variables, quantitative engagement measures, qualitative therapist observations, and administrative fields. Each feature was reviewed to determine whether it should be retained for quantitative analysis, retained for qualitative NLP analysis, or removed because it does not contribute to the research objectives.

## Removing Empty Columns

The dataset contained one unnamed column with 100% missing values. Since it contained no information and offered no analytical value, it was removed from the dataset.

In [13]:
df = df.drop(columns=["Unnamed: 19"])

In [14]:
df.shape

(512, 73)

# Missing Data Treatment Strategy

Following the missing value analysis, an appropriate strategy for handling missing data was determined. Rather than applying a single approach to all variables, missing values were handled according to the type and purpose of each feature.

This approach preserves as much useful information as possible while maintaining data quality for both the quantitative machine learning analysis and the qualitative natural language processing analysis.

## Identifying Data Types

The dataset contains both numerical and textual variables. Since these require different preprocessing techniques, the variables are separated according to their data type before further cleaning.

In [15]:
df.dtypes

Participant id                                                                                                                                                 int64
Age                                                                                                                                                            int64
Gender                                                                                                                                                        object
Special_interest                                                                                                                                              object
Any co-existing disabiltiy diagnosis                                                                                                                          object
Co-existing disability diagnosed details                                                                                                                      object
Level of S

## Identifying Object Variables

Object variables were listed to distinguish categorical and textual features from numerical variables. This inspection also helped identify variables that may require conversion to a more appropriate data type.

In [16]:
object_columns = df.select_dtypes(include="object").columns

object_columns

Index(['Gender', 'Special_interest', 'Any co-existing disabiltiy diagnosis',
       'Co-existing disability diagnosed details',
       'Stimming behaviour Identified?', 'Observed_stimming', 'Primary Goal',
       'Other information eg anymedicalcondtion selective autism',
       'Session date', 'Submitted_by', 'therapist_parent_name', 'Comment_Q1',
       'Comment_Q2', 'Comment_Q3', 'Comment_Q4', 'Comment_Q5', 'Comment_Q6',
       'Comment_Q7', 'Comment_Q8', 'Comment_Q9', 'Comment_Q10', 'Comment_Q11',
       'How did retelling the stories at home impact the carer/participant relationship?',
       'Comment_Q12', 'Comment_Q13', 'Comment_Q14',
       'What is the response time of the participant? (in minutes roughly)',
       'Comment_Q15', 'Did the response time decrease from the last session?',
       'Comment_Q16', 'Did the response time increase from the last session?',
       'Comment_Q17', 'Theme of Today's Story', 'Comment_Q18', 'Comment_Q20',
       'Comment_Q21', 'Comment_Q22', 

The object variables include a combination of demographic information, administrative fields, qualitative therapist observations, and variables that may contain numerical values stored as text.

Each group will require a different preprocessing strategy to ensure that information is preserved while preparing the dataset for both clustering and natural language processing.

## Inspecting Potential Numeric Variables

Some variables were stored as text despite appearing to represent numerical values. These variables were inspected to determine whether they could be converted into numeric data types for quantitative analysis.

In [17]:
df["What is the response time of the participant? (in minutes roughly)"].value_counts(dropna=False)

What is the response time of the participant? (in minutes roughly)
NaN                       207
1 Minutes                  64
2 Minutes                  32
30 seconds                 29
5 Seconds                  20
60 seconds                 20
3 Minutes                  14
1 Minute                   10
20 seconds                  9
5 Minutes                   9
1 Sec                       8
2 Seconds                   8
10 seconds                  7
10 Seconds                  5
1 Minutes 30 Sec            5
4 seconds                   5
Did not answer              5
15 sec                      4
120 seconds                 4
5 seconds                   3
Not response                3
1 Mintues                   3
5 second                    3
3 Second                    3
4 Minutes                   2
3-4  minutes                2
4-5 minutes                 2
10 second                   2
3 minutes                   2
3 secs                      2
4 minutes                   2
45 

## Inspecting Response Time

The response time variable was expected to contain numerical values representing the participant's response latency. However, inspection revealed inconsistent formatting, including different units (seconds and minutes), spelling variations, ranges, and descriptive text. Therefore, additional cleaning and standardisation are required before the variable can be used for quantitative analysis.

In [18]:
sorted(df["What is the response time of the participant? (in minutes roughly)"]
       .dropna()
       .unique())

['1  Minute',
 '1 Minitues',
 '1 Mintues',
 '1 Minute',
 '1 Minutes',
 '1 Minutes ',
 '1 Minutes 30 Sec',
 '1 Minutes 30 Seconds',
 '1 Sec',
 '1 minutes',
 '10 Seconds',
 '10 second',
 '10 seconds',
 '120 seconds',
 '15 sec',
 '2 Mintues',
 '2 Minutes',
 '2 Seconds',
 '2 second',
 '2 seond',
 '20 Seconds',
 '20 seconds',
 '25 Seconds',
 '3 Minutes',
 '3 Minutes 30 Seconds',
 '3 Second',
 '3 minutes',
 '3 secs',
 '3-4  minutes',
 '30 Seconds',
 '30 seconds',
 '4 Minutes',
 '4 minutes',
 '4 seconds',
 '4-5 minutes',
 '45 Sec',
 '5 Minutes',
 '5 Second',
 '5 Seconds',
 '5 second',
 '5 seconds',
 '60 seconds',
 'About 10 to 20 seconds',
 'Did not answer',
 'He did not response',
 'Just said hello',
 'Not response',
 'Nothing']

## Standardising the Response Time Variable

Inspection of the response time variable revealed inconsistent formatting, including variations in spelling, capitalisation, units of measurement, and descriptive text.

Before converting the values into a numerical format, the text was standardised to improve consistency and facilitate subsequent data transformation.

### Standardising Common Spelling Variations

Following text standardisation, several spelling inconsistencies remained (e.g., "mintues", "minitues", and "seond"). These were corrected to ensure that equivalent response times were represented consistently before numerical conversion.

In [19]:
response_col = "What is the response time of the participant? (in minutes roughly)"

df[response_col] = (
    df[response_col]
    .replace({
        "1 mintues": "1 minutes",
        "2 mintues": "2 minutes",
        "1 minitues": "1 minutes",
        "2 seond": "2 second"
    })
)

In [20]:
response_col = "What is the response time of the participant? (in minutes roughly)"

In [21]:
sorted(df[response_col].dropna().unique())

['1  Minute',
 '1 Minitues',
 '1 Mintues',
 '1 Minute',
 '1 Minutes',
 '1 Minutes ',
 '1 Minutes 30 Sec',
 '1 Minutes 30 Seconds',
 '1 Sec',
 '1 minutes',
 '10 Seconds',
 '10 second',
 '10 seconds',
 '120 seconds',
 '15 sec',
 '2 Mintues',
 '2 Minutes',
 '2 Seconds',
 '2 second',
 '20 Seconds',
 '20 seconds',
 '25 Seconds',
 '3 Minutes',
 '3 Minutes 30 Seconds',
 '3 Second',
 '3 minutes',
 '3 secs',
 '3-4  minutes',
 '30 Seconds',
 '30 seconds',
 '4 Minutes',
 '4 minutes',
 '4 seconds',
 '4-5 minutes',
 '45 Sec',
 '5 Minutes',
 '5 Second',
 '5 Seconds',
 '5 second',
 '5 seconds',
 '60 seconds',
 'About 10 to 20 seconds',
 'Did not answer',
 'He did not response',
 'Just said hello',
 'Not response',
 'Nothing']

In [22]:
df[response_col].dropna().head(10)

17    2 Minutes
18    2 Minutes
19    2 Minutes
20       15 sec
21       15 sec
22       15 sec
23       15 sec
24    5 Minutes
25    3 Minutes
26    5 Minutes
Name: What is the response time of the participant? (in minutes roughly), dtype: object

In [23]:
df[response_col] = (
    df[response_col]
    .str.lower()
    .str.strip()
)

In [24]:
df[response_col].dropna().head(10)

17    2 minutes
18    2 minutes
19    2 minutes
20       15 sec
21       15 sec
22       15 sec
23       15 sec
24    5 minutes
25    3 minutes
26    5 minutes
Name: What is the response time of the participant? (in minutes roughly), dtype: object

In [25]:
sorted(df[response_col].dropna().unique())

['1  minute',
 '1 minitues',
 '1 mintues',
 '1 minute',
 '1 minutes',
 '1 minutes 30 sec',
 '1 minutes 30 seconds',
 '1 sec',
 '10 second',
 '10 seconds',
 '120 seconds',
 '15 sec',
 '2 mintues',
 '2 minutes',
 '2 second',
 '2 seconds',
 '20 seconds',
 '25 seconds',
 '3 minutes',
 '3 minutes 30 seconds',
 '3 second',
 '3 secs',
 '3-4  minutes',
 '30 seconds',
 '4 minutes',
 '4 seconds',
 '4-5 minutes',
 '45 sec',
 '5 minutes',
 '5 second',
 '5 seconds',
 '60 seconds',
 'about 10 to 20 seconds',
 'did not answer',
 'he did not response',
 'just said hello',
 'not response',
 'nothing']

### Converting Response Time Values into Seconds

After standardising the text, valid response-time observations were converted into a single numerical unit: seconds. Using one consistent unit avoids ambiguity and makes the variable suitable for statistical analysis and machine-learning models.

For approximate ranges, the midpoint was used. For example, `3–4 minutes` was represented as 210 seconds. Textual non-responses were treated as missing values because they did not provide a measurable response time.

In [28]:
def convert_response_time_to_seconds(value):
    if pd.isna(value):
        return np.nan

    value = str(value).lower().strip()

    # Correct common spelling variations
    value = (
        value
        .replace("mintues", "minutes")
        .replace("minitues", "minutes")
    )

    non_response_values = {
        "did not answer",
        "he did not response",
        "just said hello",
        "not response",
        "nothing"
    }

    if value in non_response_values:
        return np.nan

    # Approximate ranges: use midpoint
    if value == "3-4  minutes":
        return 3.5 * 60

    if value == "4-5 minutes":
        return 4.5 * 60

    if value == "about 10 to 20 seconds":
        return 15

    minute_match = re.search(r"(\d+(?:\.\d+)?)\s*minute", value)
    minutes = float(minute_match.group(1)) if minute_match else 0

    second_match = re.search(r"(\d+(?:\.\d+)?)\s*(?:second|sec)", value)
    seconds = float(second_match.group(1)) if second_match else 0

    total_seconds = (minutes * 60) + seconds

    return total_seconds if total_seconds > 0 else np.nan

In [30]:
import re
df["response_time_seconds"] = df[response_col].apply(
    convert_response_time_to_seconds
)

In [31]:
df[[response_col, "response_time_seconds"]].dropna().head(20)

,What is the response time of the participant? (in minutes roughly),response_time_seconds
17,2 minutes,120.0
18,2 minutes,120.0
19,2 minutes,120.0
20,15 sec,15.0
21,15 sec,15.0
22,15 sec,15.0
23,15 sec,15.0
24,5 minutes,300.0
25,3 minutes,180.0
26,5 minutes,300.0


In [32]:
df["response_time_seconds"].describe()

count    294.000000
mean      69.159864
std       69.436923
min        1.000000
25%       16.250000
50%       60.000000
75%       90.000000
max      300.000000
Name: response_time_seconds, dtype: float64

In [33]:
conversion_check = df[
    df[response_col].notna() &
    df["response_time_seconds"].isna()
][response_col].value_counts()

conversion_check

What is the response time of the participant? (in minutes roughly)
did not answer         5
not response           3
he did not response    1
nothing                1
just said hello        1
Name: count, dtype: int64

### Final Validation of the Response-Time Variable

The converted response-time variable was validated by identifying records that contained an original response-time entry but no numerical value after conversion.

Inspection confirmed that all remaining missing values represented genuine non-response observations (e.g., "did not answer" and "not response") rather than conversion errors. Therefore, the preprocessing procedure successfully transformed all measurable response times into a consistent numerical variable expressed in seconds.

##✅ Response Time Status: COMPLETE

You've now successfully cleaned one of the messiest variables in the datase

In [34]:
df["Engagement_opportunities_count"].value_counts(dropna=False)

Engagement_opportunities_count
10                  144
NaN                 130
5                    47
4                    37
6                    36
3                    21
8                    21
7                    17
9                    15
12                    9
Data not cleared      8
2                     6
1                     5
16                    4
15                    3
14                    2
4 to 5                2
5 to 6                2
40                    1
No data               1
11                    1
Name: count, dtype: int64

In [35]:
engagement_col = "Engagement_opportunities_count"

In [36]:
df[engagement_col].dtype

dtype('O')

In [43]:
df[engagement_col] = df[engagement_col].replace({
    "Data not cleared": np.nan,
    "No data": np.nan
})

In [44]:
df[engagement_col] = df[engagement_col].replace({
    "4 to 5": 4.5,
    "5 to 6": 5.5
})

In [46]:
df[df[engagement_col].astype(str).str.contains("data", case=False, na=False)][engagement_col]

Series([], Name: Engagement_opportunities_count, dtype: object)

In [47]:
df[engagement_col] = df[engagement_col].apply(
    lambda x: x.strip() if isinstance(x, str) else x
)

In [48]:
df[engagement_col] = df[engagement_col].replace({
    "Data not cleared": np.nan,
    "No data": np.nan
})

In [49]:
df[engagement_col].unique()

array(['10', '3', '8', '4', '2', nan, '5', '6', '12', '7', '9', '40',
       '15', '1', '16', '4 to 5', '5 to 6', '14', '11'], dtype=object)

In [50]:
df[engagement_col] = df[engagement_col].replace({
    "4 to 5": 4.5,
    "5 to 6": 5.5
})

In [51]:
df[engagement_col].unique()

array(['10', '3', '8', '4', '2', nan, '5', '6', '12', '7', '9', '40',
       '15', '1', '16', 4.5, 5.5, '14', '11'], dtype=object)

In [54]:
df[engagement_col] = pd.to_numeric(df[engagement_col])

In [56]:
df[engagement_col].dtype

dtype('float64')

In [57]:
df[engagement_col].describe()

count    373.000000
mean       7.659517
std        3.438695
min        1.000000
25%        5.000000
50%        8.000000
75%       10.000000
max       40.000000
Name: Engagement_opportunities_count, dtype: float64

### Validation of the Engagement Opportunities Count Variable



Following preprocessing, the variable was successfully converted into a numerical format. Descriptive statistics were examined to verify that the conversion had been completed correctly and to identify any potential outliers or unexpected values requiring further investigation.

### Initial Validation

Descriptive statistics indicated that the cleaned variable had 373 valid observations. Most engagement opportunity counts ranged between 5 and 10, although one observation recorded a value of 40. This observation was retained at this stage and will be investigated later during the outlier analysis rather than being removed without evidence.

In [58]:
df[df[engagement_col] == 40]

,Participant id,Age,Gender,Special_interest,Any co-existing disabiltiy diagnosis,Co-existing disability diagnosed details,Level of Severity,Autism Level,Stimming behaviour Identified?,Observed_stimming,Primary Goal,Other information eg anymedicalcondtion selective autism,Session number,Session date,Submitted_by,therapist_parent_name,How engaged was the participant during today's storytelling session?,Comment_Q1,How well did the participant understand that this story is personalised for them?,Comment_Q2,Did the participant demonstrate emotional connection?,Comment_Q3,How would you rate the participant’s verbal participation?,Comment_Q4,Did the participant maintain attention throughout the session?,Comment_Q5,"How likely a participant was to retell and use the stories in his/her daily conversation or interaction with\ntheir peers, carers, parents or therapists?",Comment_Q6,Did the participant show any signs of enjoyment during the session?,Comment_Q7,"Did the participant exhibit distress, boredom, or frustration?",Comment_Q8,Did the participant initiate interactions related to the story?,Comment_Q9,Did the participant repeat the story or any content from the story?,Comment_Q10,Did the participant want to or try to creatively change the story?,Comment_Q11,How did retelling the stories at home impact the carer/participant relationship?,Comment_Q12,How much did the relationship between a participant and carer/parent improved?,Comment_Q13,Did the participant express their feelings about the story (verbally or otherwise)?,Comment_Q14,What is the response time of the participant? (in minutes roughly),Comment_Q15,Did the response time decrease from the last session?,Comment_Q16,Did the response time increase from the last session?,Comment_Q17,Theme of Today's Story,To what extent did the participant understand the story theme?,Comment_Q18,Did the participant apply the learning during the session?,Comment_Q20,Do you feel that the participant is confident and has the potential to be able to apply this story in the real world after a session?,Comment_Q21,Did the participant generalise the behaviour outside the story?,Comment_Q22,Did the participant recall or refer to a previous story or theme?,Comment_Q23,Did the participant reflect on or comment about the theme after the story ended?,Comment_Q24,Did the participant link the story to real-life experiences?,Comment_Q25,How much different scenarios stories impact overall social behaviour ?,Comment_Q26,Theme_specific_situation,Engagement_opportunities_count,Successful/postive response,Success_percentage,Notes_intervention,Additional_notes_observations,response_time_seconds
129,111,16,Male,"Food such as fast food,grilled chicken . Vehic...",Yes,He has attention deficts disorder,1,1,Yes,Repeatedly asking the same question is one of ...,NaN,He has depressive disorder with associated agg...,3,NaN,T,Asad,2.0,NaN,3.0,NaN,3.0,NaN,7.0,NaN,3.0,NaN,3.0,NaN,3.0,NaN,2.0,NaN,3.0,NaN,2.0,NaN,2.0,NaN,NaN,NaN,2.0,NaN,2.0,NaN,NaN,NaN,No,NaN,No,NaN,NaN,3.0,NaN,2.0,NaN,3.0,NaN,3.0,NaN,3.0,NaN,3.0,NaN,3.0,NaN,5,NaN,Story of Bird,40.0,30,75%,NaN,"While in the session setting, his performance ...",NaN


### Investigation of Potential Outlier

One observation recorded 40 engagement opportunities, which was substantially higher than the majority of observations. This record was investigated by comparing it with the corresponding number of successful responses and success percentage. The values were internally consistent (40 opportunities, 30 successful responses, 75% success), indicating that the observation was plausible. Therefore, the value was retained in the dataset.

In [59]:
df.dtypes[df.dtypes == "object"]

Gender                                                                              object
Special_interest                                                                    object
Any co-existing disabiltiy diagnosis                                                object
Co-existing disability diagnosed details                                            object
Stimming behaviour Identified?                                                      object
Observed_stimming                                                                   object
Primary Goal                                                                        object
Other information eg anymedicalcondtion selective autism                            object
Session date                                                                        object
Submitted_by                                                                        object
therapist_parent_name                                                               object

In [60]:
df["Successful/postive response"].value_counts(dropna=False)

Successful/postive response
NaN                 128
4                    50
6                    41
2                    37
5                    37
8                    35
9                    34
3                    32
10                   28
1                    26
7                    25
0                    12
Data not cleared      8
12                    4
15                    4
16                    3
5 to 6                2
4 to 5                2
30                    1
No data               1
14                    1
24                    1
Name: count, dtype: int64

In [61]:
success_col = "Successful/postive response"

In [62]:
df[success_col] = df[success_col].apply(
    lambda x: x.strip() if isinstance(x, str) else x
)

df[success_col] = df[success_col].replace({
    "Data not cleared": np.nan,
    "No data": np.nan
})

In [63]:
df[success_col] = df[success_col].replace({
    "4 to 5": 4.5,
    "5 to 6": 5.5
})

In [64]:
df[success_col].unique()

array(['3', '5', '1', '4', '2', nan, '0', '8', '10', '6', '7', '9', '12',
       '30', '15', '16', 4.5, 5.5, '14', '24'], dtype=object)

In [65]:
df[success_col] = pd.to_numeric(df[success_col])

In [66]:
df[success_col].dtype

dtype('float64')

In [67]:
df[success_col].describe()

count    375.000000
mean       5.632000
std        3.584062
min        0.000000
25%        3.000000
50%        5.000000
75%        8.000000
max       30.000000
Name: Successful/postive response, dtype: float64

In [68]:
df.loc[
    df[success_col] >= 24,
    [
        "Participant id",
        "Session number",
        "Engagement_opportunities_count",
        success_col,
        "Success_percentage"
    ]
]

,Participant id,Session number,Engagement_opportunities_count,Successful/postive response,Success_percentage
129,111,3,40.0,30.0,75%
433,132,3,1.0,24.0,20%


In [70]:
df.loc[433]

Participant id                                                                                                                                                                                             132
Age                                                                                                                                                                                                          9
Gender                                                                                                                                                                                                    Male
Special_interest                                                                                                                                             Foods such as Pizza, interest in ducklings and...
Any co-existing disabiltiy diagnosis                                                                                                                                        

### Investigation of an Inconsistent Observation

One record contained an internally inconsistent value for the number of successful responses. The participant had only one recorded engagement opportunity but 24 successful responses, with a reported success percentage of 20%. The remaining questionnaire responses and session notes also indicated very low engagement. As the correct value could not be determined from the available information, the successful response value was treated as missing rather than being replaced with a guessed value.

In [71]:
df.loc[433, success_col] = np.nan

In [72]:
df[success_col].describe()

count    374.000000
mean       5.582888
std        3.460203
min        0.000000
25%        3.000000
50%        5.000000
75%        8.000000
max       30.000000
Name: Successful/postive response, dtype: float64

### Validation of the Successful Response Variable

The *Successful/postive response* variable was cleaned by replacing invalid text values with missing values, converting ranges to their midpoint values, and converting the column to a numeric format. During validation, one observation contained an internally inconsistent value (24 successful responses despite only one engagement opportunity). As the correct value could not be verified, this observation was treated as missing rather than being replaced with an assumed value. The cleaned variable contained 374 valid observations and was retained for subsequent analyses.

In [73]:
df["Success_percentage"].value_counts(dropna=False)

Success_percentage
NaN       150
100%       83
80%        39
50%        34
90%        26
60%        25
70%        18
40%        16
0%         12
75%        11
25%        10
30%         8
100         8
83%         7
20          7
66%         7
67%         6
20%         6
33%         6
85%         5
87.50%      3
89%         3
78%         2
86%         2
71%         2
68%         2
25          2
10%         2
77%         1
40          1
44%         1
2           1
90          1
93%         1
3%          1
27%         1
43%         1
42%         1
Name: count, dtype: int64

In [74]:
percentage_col = "Success_percentage"

In [75]:
df[percentage_col]

0       30%
1       50%
2       50%
3       30%
4       50%
       ... 
507    100%
508    100%
509    100%
510    100%
511    100%
Name: Success_percentage, Length: 512, dtype: object

In [76]:
df[df[percentage_col].notna() &
   ~df[percentage_col].astype(str).str.contains("%")][percentage_col]

193     20
194     90
195     40
196     25
200      2
304    100
305    100
306    100
307    100
308    100
309    100
310    100
311    100
312     20
313     20
314     20
315     20
316     20
317     20
318     25
Name: Success_percentage, dtype: object

In [77]:
df.loc[
    df[percentage_col].notna() &
    ~df[percentage_col].astype(str).str.contains("%"),
    [
        "Engagement_opportunities_count",
        "Successful/postive response",
        percentage_col
    ]
]

,Engagement_opportunities_count,Successful/postive response,Success_percentage
193,10.0,2.0,20
194,10.0,4.0,90
195,5.0,2.0,40
196,4.0,1.0,25
200,2.0,2.0,2
304,4.0,4.0,100
305,4.0,4.0,100
306,5.0,5.0,100
307,10.0,10.0,100
308,16.0,16.0,100


In [79]:
df["success_percentage_calculated"] = (
    df[success_col] / df[engagement_col]
) * 100

In [80]:
df[
    [
        engagement_col,
        success_col,
        percentage_col,
        "success_percentage_calculated"
    ]
].head(20)

,Engagement_opportunities_count,Successful/postive response,Success_percentage,success_percentage_calculated
0,10.0,3.0,30%,30.000000
1,10.0,5.0,50%,50.000000
2,10.0,5.0,50%,50.000000
3,3.0,1.0,30%,33.333333
4,8.0,4.0,50%,50.000000
5,4.0,2.0,50%,50.000000
6,2.0,2.0,100%,100.000000
7,NaN,NaN,NaN,NaN
8,NaN,NaN,NaN,NaN
9,NaN,NaN,NaN,NaN


### Success Percentage

The original *Success_percentage* variable contained several inconsistencies, including missing percentage symbols and values that did not correspond to the recorded engagement opportunities and successful responses. Rather than attempting to correct these manually, a new variable (`success_percentage_calculated`) was computed directly from the cleaned engagement opportunity and successful response variables. This ensured complete consistency while preserving the original variable for reference.

## 10. Save Cleaned Dataset

After completing the data cleaning process, the cleaned dataset is saved separately from the raw dataset. This ensures that the original data remain unchanged and that all subsequent analyses use the cleaned version.

In [81]:
processed_data_folder = project_root / "data" / "processed"

In [82]:
raw_data_folder = project_root / "data" / "raw"

In [83]:
list(processed_data_folder.iterdir())

[]

In [84]:
df.to_csv(
    processed_data_folder / "cleaned_data.csv",
    index=False,
    encoding="utf-8"
)

In [85]:
list(processed_data_folder.iterdir())

[WindowsPath('c:/Users/puspi/OneDrive/Desktop/Dissertation/data/processed/cleaned_data.csv')]

# Conclusion

The data cleaning process has been completed successfully. All preprocessing decisions were documented and validated before being applied. The cleaned dataset has been saved separately from the original raw dataset to preserve data integrity. Subsequent analyses, including feature engineering, clustering, natural language processing, and predictive modelling, will use the cleaned dataset stored in the `data/processed` directory.